src/EntityId/PrefixMappingEntityIdParser.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Wikibase\DataModel\Services\EntityId;
+
+use Wikimedia\Assert\Assert;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdParser;
+
+/**
+ * EntityIdParser that adds a fixed prefix and parses resulting string as an EntityId.
+ * This can be used to prefix IDs of entities coming from a foreign repository
+ * with the repository name to avoid clashes with IDs of local entities.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class PrefixMappingEntityIdParser implements EntityIdParser {
+
+	/**
+	 * @var string
+	 */
+	private $prefix;
+
+	/**
+	 * @var string[]
+	 */
+	private $prefixMapping;
+
+	/**
+	 * @var EntityIdParser
+	 */
+	private $idParser;
+
+	/**
+	 * @param string $prefix Prefix to be added. It should not contain a colon, in particular at the end of the prefix
+	 * @param string[] $prefixMapping
+	 * @param EntityIdParser $idParser
+	 */
+	public function __construct( $prefix, array $prefixMapping, EntityIdParser $idParser ) {
+		Assert::parameterType( 'string', $prefix, '$prefix' );
+		Assert::parameter( strpos( $prefix, ':' ) === false, '$prefix', 'must not contain a colon' );
+		$this->prefix = $prefix;
+		$this->prefixMapping = $prefixMapping;
+		$this->idParser = $idParser;
+	}
+
+	/**
+	 * Adds the prefix to the given string, maps the resulting string according to prefix mapping definition
+	 * and parses the resulting string as an EntityId.
+	 *
+	 * @param string $idSerialization
+	 * @return EntityId
+	 */
+	public function parse( $idSerialization ) {
+		$prefixedIdSerialization = EntityId::joinSerialization( array( $this->prefix, '', $idSerialization ) );
+		$serializationParts = EntityId::splitSerialization( $idSerialization );
+		$serializationRepoName = $serializationParts[0];
+		if ( array_key_exists( $serializationRepoName, $this->prefixMapping ) ) {
```

Comment:
I generally prefer isset( $this->prefixMapping[$serializationRepoName] ) for readability. Note that these behave slightly differently: isset() will return falls if the key exists but the value is null, while array_key_exists() will return true in that case. That difference isn't relevant here, though.